xcube/util/types.py

```diff
@@ -0,0 +1,75 @@
+# The MIT License (MIT)
+# Copyright (c) 2022 by the xcube development team and contributors
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from typing import Optional, Tuple, Type, TypeVar, Union
+
+from xcube.util.assertions import assert_true
+
+T = TypeVar('T', int, float)
+
+ItemType = Union[Type[T], Tuple[Type[T], ...]]
+Pair = Tuple[T, T]
+ScalarOrPair = Union[T, Pair]
+
+
+def normalize_scalar_or_pair(
+        value: Optional[ScalarOrPair] = None,
+        *,
+        item_type: Optional[ItemType[T]] = None,
+        default: Optional[ScalarOrPair] = None,
```

Comment:
To me, a `default` value does not make sense, because the function is about _normalization_ of `value`.
In addition, it is fine for `value` to be `None` as a value, not as an indicator for absense. The result would be `(None, None)` which may be desired.

Reply:
> To me, a `default` value does not make sense, because the function is about _normalization_ of `value`. 

Again, this is in line with existing normalization behaviour (same link as above): https://github.com/dcs4cop/xcube/blob/d3fc869ce1a80541d876cb1a0b484b2bb5956dd0/xcube/core/gridmapping/helpers.py#L89

There would be uses for a helper function that applies a default if the normalization value is not valid. Maybe this could be a second function?

> In addition, it is fine for `value` to be `None` as a value, not as an indicator for absense. The result would be `(None, None)` which may be desired.

Agreed.